lib/autoload/duplicator.rb

```diff
@@ -0,0 +1,26 @@
+class Duplicator
+  # Create an instance of Duplicator to track duplicated objects.
+  #
+  # @param [Array] excluded_objects A list of objects to be excluded from duplication
+  def initialize(excluded_objects=[])
+    @duplicated_objects = {}  # hash to check what has been duplicated
+    @exclusion_set = excluded_objects.to_set  # set to check what should be excluded
+  end
+
+  # Deep copy source_object and its children. source_object must provide a duplicate
+  # method which duplicates its children.
+  #
+  # @param [Object#duplicate] source_object The object to be duplicated.
+  # @return duplicated_object A reference to the duplicated object.
+  def duplicate_object(source_object)
```

Comment:
There are 2 objects in the lookup hash.
`a.children[0].children[0] is a`

The `expect` statements cover some of these checks. I've also used `inspect` to look at the object IDs to make sure they're the same object.

Is there something I'm missing?